consumer/consumererror/permanent.go

```diff
@@ -14,27 +14,50 @@
 
 package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
 
-import "errors"
+import (
+	"errors"
+	"strconv"
+)
 
-// permanent is an error that will be always returned if its source
+// noDataCount indicates an unknown number of records.
+const noDataCount = -1
+
+// Permanent is an error that will always be returned if its source
 // receives the same inputs.
-type permanent struct {
-	err error
+type Permanent struct {
+	error
+	rejected int
 }
 
-// NewPermanent wraps an error to indicate that it is a permanent error, i.e. an
-// error that will be always returned if its source receives the same inputs.
+// Deprecated [v0.76.0] Use consumererror.NewPermanentWithCount instead.
 func NewPermanent(err error) error {
-	return permanent{err: err}
+	return Permanent{error: err, rejected: noDataCount}
+}
+
+// NewPermanentWithCount wraps an error to indicate that it is a permanent error, i.e. an
+// error that will be always returned if its source receives the same inputs.
+// The error should also include the number of rejected records.
+func NewPermanentWithCount(err error, rejected int) error {
```

Comment:
I'm a bit worried that we require providing the rejected field here. I'm wondering if we can keep this field optional. If not provided, the caller can assume that all the items got rejected. My concern is that some exporters return this error when there is no access to the original pdata that was translated to some other schema. It'll be difficult for them to fetch the number of the items. One option is to have
```
NewPermanent(err error, otps ...PermanentErrorOptions)
```

and supply `rejected` via `WithRejectedCount(int)` option.

Or we don't even need `Permanent` part here and assume it's always permanent unless `Retryable(ptrace.Traces)` option is provided. Later we might have some other options.

These are just thoughts, not suggestions for changes. I believe we need to think more about an overarching design for this API that would cover all the use cases, including embedding the GRPC/HTTP and possible future additions. Maybe having some small design doc would be beneficial.

Reply:
I agree that having a design doc would be good, I'll think through some of your points and add them to a design doc here. If it looks like the PR is growing we can move the document to a separate PR.